json_object.c

```diff
@@ -720,24 +755,116 @@ int64_t json_object_get_int64(const struct json_object *jso)
+	{
+	case json_type_int:
+		switch(jso->o.c_int.cint_type) {
+		case json_object_int_type_int64:
+			if (jso->o.c_int.cint.c_int64 < 0)
+				return 0;
+			return (uint64_t)jso->o.c_int.cint.c_int64;
+		case json_object_int_type_uint64:
+			return jso->o.c_int.cint.c_uint64;
+		default:
+			return 0;
+		}
+	case json_type_double:
+		// UINT64_MAX can't be exactly represented as a double
+		// so cast to tell the compiler it's ok to round up.
+		if (jso->o.c_double >= (double)UINT64_MAX)
+			return UINT64_MAX;
+		if (jso->o.c_double < 0)
+			return 0;
+		return (uint64_t)jso->o.c_double;
+	case json_type_boolean:
+		return jso->o.c_boolean;
+	case json_type_string:
+		if (json_parse_uint64(get_string_component(jso), &cuint) == 0)
+			return cuint;
+		/* FALLTHRU */
+	default:
+		return 0;
+	}
+}
+
 int json_object_set_int64(struct json_object *jso,int64_t new_value){
 	if (!jso || jso->o_type!=json_type_int)
 		return 0;
-	jso->o.c_int64=new_value;
+	jso->o.c_int.cint.c_int64=new_value;
+	jso->o.c_int.cint_type = json_object_int_type_int64;
+	return 1;
+}
+
+int json_object_set_uint64(struct json_object *jso,uint64_t new_value){
+	if (!jso || jso->o_type!=json_type_int)
+		return 0;
+	jso->o.c_int.cint.c_uint64=new_value;
+	jso->o.c_int.cint_type = json_object_int_type_uint64;
 	return 1;
 }
 
 int json_object_int_inc(struct json_object *jso, int64_t val) {
 	if (!jso || jso->o_type != json_type_int)
 		return 0;
-	if (val > 0 && jso->o.c_int64 > INT64_MAX - val) {
-		jso->o.c_int64 = INT64_MAX;
-	} else if (val < 0 && jso->o.c_int64 < INT64_MIN - val) {
-		jso->o.c_int64 = INT64_MIN;
-	} else {
-		jso->o.c_int64 += val;
+	switch(jso->o.c_int.cint_type) {
+	case json_object_int_type_int64:
+		if (val > 0 && jso->o.c_int.cint.c_int64 > INT64_MAX - val) {
+			jso->o.c_int.cint.c_uint64 = (uint64_t)jso->o.c_int.cint.c_int64 + (uint64_t)val;
+			jso->o.c_int.cint_type = json_object_int_type_uint64;
+		} else if (val < 0 && jso->o.c_int.cint.c_int64 < INT64_MIN - val) {
+			jso->o.c_int.cint.c_int64 = INT64_MIN;
+		} else {
+			jso->o.c_int.cint.c_int64 += val;
+		}
+		return 1;
+	case json_object_int_type_uint64:
+		if (val > 0 && jso->o.c_int.cint.c_uint64 > UINT64_MAX - (uint64_t)val) {
+			jso->o.c_int.cint.c_uint64 = UINT64_MAX;
+		} else if (val < 0 && jso->o.c_int.cint.c_uint64 < (uint64_t)(-val)) {
+			jso->o.c_int.cint.c_int64 = (int64_t)jso->o.c_int.cint.c_uint64 + val;
+			jso->o.c_int.cint_type = json_object_int_type_int64;
+		} else if (val < 0 && jso->o.c_int.cint.c_uint64 >= (uint64_t)(-val)) {
+			jso->o.c_int.cint.c_uint64 = jso->o.c_int.cint.c_uint64 - (uint64_t)(-val);
+		} else {
+			jso->o.c_int.cint.c_uint64 += val;
+		}
+		return 1;
+	default:
+		return 0;
+	}
+}
+
+int json_object_uint_inc(struct json_object *jso, uint64_t val) {
+	if (!jso || jso->o_type != json_type_int)
+		return 0;
+	switch(jso->o.c_int.cint_type) {
+	case json_object_int_type_int64:
+		if ((uint64_t)jso->o.c_int.cint.c_uint64 + val> UINT64_MAX) {
```

Comment:
This can't ever be true.  There are all kinds of overflow pitfalls here.  Also, I'm not sure it's even worth having a separate json_object_uint_inc() at all, but even if it is I feel like it should be possible to use some common code between this and json_object_int_inc().

Reply:
ha, It's my fault. Should be c_int64. There is some same errors here.

Reply:
Or delete json_object_uint_inc（）?

Reply:
It's also likely to be error prone in that passing a negative value for val won't do what you might expect a first glance.
Let's drop this function, at least for now, and only consider adding it if someone has an actual need for it.

Reply:
OK，I will delete it. If need it , I will add ```json_object_uint_inc()``` later.